kernel/src/checkpoint/tests.rs

```diff
@@ -292,10 +382,54 @@ fn test_v1_checkpoint_specific_version() -> DeltaResult<()> {
     // No more data should exist because we only requested version 0
     assert!(data_iter.next().is_none());
 
-    assert_eq!(data_iter.actions_count, 2);
-    assert_eq!(data_iter.add_actions_count, 0);
+    // Finalize and verify checkpoint metadata
+    let size_in_bytes = 10;
+    let metadata = FileMeta {
+        location: Url::parse("memory:///fake_path_2")?,
+        last_modified: 0,
+        size: size_in_bytes,
+    };
+    writer.finalize(&engine, &metadata, data_iter)?;
+    assert_last_checkpoint_contents(
+        &store,
+        0,                                 // version: specified version (0)
+        2,                                 // size: 1 protocol + 1 metadata from version 0
+        0,                                 // numOfAddFiles: no add files in version 0
+        size_in_bytes.try_into().unwrap(), // sizeInBytes: passed to finalize (10)
+    )?;
+
+    Ok(())
+}
+
+#[test]
+fn test_finalize_errors_if_checkpoint_data_iterator_is_not_exhausted() -> DeltaResult<()> {
+    let (store, _) = new_in_memory_store();
+    let engine = DefaultEngine::new(store.clone(), Arc::new(TokioBackgroundExecutor::new()));
+
+    // 1st commit (version 0) - metadata and protocol actions
+    write_commit_to_store(
+        &store,
+        vec![create_basic_protocol_action(), create_metadata_action()],
+        0,
+    )?;
+
+    let table_root = Url::parse("memory:///")?;
+    let table = Table::new(table_root);
+    let writer = table.checkpoint(&engine, Some(0))?;
+    let data_iter = writer.checkpoint_data(&engine)?;
+
+    /* The returned data iterator has batches that we do not consume */
+
+    let size_in_bytes = 10;
+    let metadata = FileMeta {
+        location: Url::parse("memory:///fake_path_2")?,
+        last_modified: 0,
+        size: size_in_bytes,
+    };
+
+    // Attempt to finalize the checkpoint with an iterator that has not been fully consumed
+    assert!(writer.finalize(&engine, &metadata, data_iter).is_err());
```

Comment:
nit (and i'm guity of this and trying to get better!) can we check actual error returned here?